modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PageMemoryPartitionStorage.java

```diff
@@ -0,0 +1,512 @@
+                            return dataRow;
+                        }
+                    }
+
+                    return null;
+                }
+            };
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Error while scanning rows", e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<Void> snapshot(Path snapshotPath) {
+        throw new UnsupportedOperationException("Snapshots are not supported yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void restoreSnapshot(Path snapshotPath) {
+        throw new UnsupportedOperationException("Snapshots are not supported yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void destroy() throws StorageException {
+        try {
+            tree.destroy();
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Error while destroying data", e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void close() {
+        tree.close();
+    }
+
+    private static TableSearchRow wrap(SearchRow searchRow) {
+        ByteBuffer key = searchRow.key();
+
+        return new TableSearchRow(StorageUtils.hashCode(key), key);
+    }
+
+    private static TableDataRow wrap(DataRow dataRow) {
+        ByteBuffer key = dataRow.key();
+        ByteBuffer value = dataRow.value();
+
+        return new TableDataRow(StorageUtils.hashCode(key), key, value);
+    }
+
+    private static @Nullable DataRow wrap(TableDataRow tableDataRow) {
+        return tableDataRow == null ? null : new TableDataRowAdapter(tableDataRow);
+    }
+
+    private static class InsertClosure implements IgniteTree.InvokeClosure<TableDataRow> {
+        final TableFreeList freeList;
+
+        TableDataRow newRow;
+
+        @Nullable TableDataRow oldRow;
+
+        InsertClosure(TableFreeList freeList) {
+            this.freeList = freeList;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public void call(@Nullable TableDataRow oldRow) {
+            this.oldRow = oldRow;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public @Nullable TableDataRow newRow() {
+            assert newRow != null;
+
+            try {
+                freeList.insertDataRow(newRow);
+            } catch (IgniteInternalCheckedException e) {
+                throw new IgniteInternalException(e);
+            }
+
+            return newRow;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public IgniteTree.@Nullable OperationType operationType() {
+            return oldRow == null ? IgniteTree.OperationType.PUT : IgniteTree.OperationType.NOOP;
+        }
+
+        void reset() {
+            newRow = null;
+
+            oldRow = null;
+        }
+    }
+
+    private static class RemoveExactClosure implements IgniteTree.InvokeClosure<TableDataRow> {
+        TableDataRow removedRow;
```

Comment:
It is not removed, it's intended for removal, right?

Reply:
Fix it